homeassistant/components/tplink_omada/update.py

```diff
@@ -0,0 +1,149 @@
+"""Support for TPLink Omada device toggle options."""
+from __future__ import annotations
+
+import logging
+from typing import Any, NamedTuple
+
+from tplink_omada_client.devices import OmadaFirmwareUpdate, OmadaListDevice
+from tplink_omada_client.omadasiteclient import OmadaSiteClient
+
+from homeassistant.components.update import UpdateEntity, UpdateEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.event import async_call_later
+
+from .const import DOMAIN
+from .controller import OmadaSiteController
+from .coordinator import OmadaCoordinator
+from .entity import OmadaDeviceEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FirmwareUpdateStatus(NamedTuple):
+    """Firmware update information for Omada SDN devices."""
+
+    device: OmadaListDevice
+    firmware: OmadaFirmwareUpdate | None
+
+
+async def _get_firmware_updates(client: OmadaSiteClient) -> list[FirmwareUpdateStatus]:
+    devices = await client.get_devices()
+    return [
+        FirmwareUpdateStatus(
+            device=d,
+            firmware=None
+            if not d.need_upgrade
+            else await client.get_firmware_details(d),
+        )
+        for d in devices
+    ]
+
+
+async def _poll_firmware_updates(
+    client: OmadaSiteClient,
+) -> dict[str, FirmwareUpdateStatus]:
+    """Poll the state of Omada Devices firmware update availability."""
+    return {d.device.mac: d for d in await _get_firmware_updates(client)}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up switches."""
+    controller: OmadaSiteController = hass.data[DOMAIN][config_entry.entry_id]
+    omada_client = controller.omada_client
+
+    devices = await omada_client.get_devices()
+
+    coordinator = OmadaCoordinator[FirmwareUpdateStatus](
+        hass,
+        omada_client,
+        "Firmware Updates",
+        _poll_firmware_updates,
+        poll_delay=6 * 60 * 60,
+    )
+
+    entities: list = []
+    for device in devices:
+        entities.append(OmadaDeviceUpdate(coordinator, device))
+
+    async_add_entities(entities)
```

Comment:
Instead of creating a temporary list, pass a generate to `async_add_entites` using a list comprehension.